client/exec.go

```diff
@@ -0,0 +1,305 @@
+		Width       int               `json:"width"`
+		Height      int               `json:"height"`
+	}{
+		Command:     opts.Command,
+		Environment: opts.Environment,
+		WorkingDir:  opts.WorkingDir,
+		Timeout:     timeoutStr,
+		UserID:      opts.UserID,
+		User:        opts.User,
+		GroupID:     opts.GroupID,
+		Group:       opts.Group,
+		Terminal:    opts.Terminal,
+		Stderr:      opts.Stderr,
+		Width:       opts.Width,
+		Height:      opts.Height,
+	}
+	var body bytes.Buffer
+	err := json.NewEncoder(&body).Encode(&payload)
+	if err != nil {
+		return "", err
+	}
+	headers := map[string]string{
+		"Content-Type": "application/json",
+	}
+	resultBytes, changeID, err := client.doAsyncFull("POST", "/v1/exec", nil, headers, &body)
+	if err != nil {
+		return "", err
+	}
+	var result struct {
+		WebsocketIDs map[string]string `json:"websocket-ids"`
+	}
+	err = json.Unmarshal(resultBytes, &result)
+	if err != nil {
+		return "", err
+	}
+
+	if args == nil {
+		return changeID, nil
+	}
+
+	// Process additional arguments (connecting I/O and websockets)
+	fds := result.WebsocketIDs
+
+	// Call the control handler with a connection to the control socket
+	if args.Control != nil && fds["control"] != "" {
+		conn, err := client.getChangeWebsocket(changeID, fds["control"])
+		if err != nil {
+			return "", err
+		}
+		go args.Control(conn)
+	}
+
+	if opts.Terminal {
+		// Handle terminal-based executions
+		if args.Stdin != nil && args.Stdout != nil {
+			// Connect to the websocket
+			conn, err := client.getChangeWebsocket(changeID, fds["io"])
+			if err != nil {
+				return "", err
+			}
+
+			// And attach stdin and stdout to it
+			go func() {
+				wsutil.WebsocketSendStream(conn, args.Stdin, -1)
+				<-wsutil.WebsocketRecvStream(args.Stdout, conn)
+				conn.Close()
+				if args.DataDone != nil {
+					close(args.DataDone)
+				}
+			}()
+		} else {
+			if args.DataDone != nil {
+				close(args.DataDone)
+			}
+		}
+	} else {
+		// Handle non-terminal executions
+		dones := map[string]chan bool{}
+		conns := []*websocket.Conn{}
+
+		// Handle stdin and stdout
+		if fds["io"] != "" {
+			conn, err := client.getChangeWebsocket(changeID, fds["io"])
+			if err != nil {
+				return "", err
+			}
+			conns = append(conns, conn)
+			dones["stdin"] = wsutil.WebsocketSendStream(conn, args.Stdin, -1)
+			dones["stdout"] = wsutil.WebsocketRecvStream(args.Stdout, conn)
+		}
+
+		// Handle stderr separately if needed
+		if opts.Stderr && fds["stderr"] != "" {
+			conn, err := client.getChangeWebsocket(changeID, fds["stderr"])
+			if err != nil {
+				return "", err
+			}
+			conns = append(conns, conn)
+			dones["stderr"] = wsutil.WebsocketRecvStream(args.Stderr, conn)
+		}
+
+		// Wait for everything to be done
+		go func() {
+			for name, done := range dones {
+				// Skip stdin, dealing with it separately below
+				if name == "stdin" {
+					continue
+				}
+				<-done
+			}
+
+			if fds["io"] != "" {
+				if args.Stdin != nil {
+					args.Stdin.Close()
```

Comment:
Again, not entirely sure about what's the overall API design that is being suggested, but in general we don't close someone else's files unless we're taking ownership of the data completely. That tends not to be the case with exec logic.

Reply:
Fixed. I'm unsure why the LXD code was closing stdin, but I don't think that's ever necessary. It's just an `io.Reader` now.